src/Chainweb/Version.hs

```diff
@@ -209,109 +206,76 @@ instance IsMerkleLogEntry ChainwebHashTag ChainwebVersion where
     {-# INLINE fromText #-}
 
+chainwebVersions :: HM.HashMap T.Text ChainwebVersion
+chainwebVersions = HM.fromList $
+    f Test "test"
+    <> f TestWithTime "testWithTime"
+    <> f TestWithPow "testWithPow"
+    <> [ ("testnet00", Testnet00) ]
+  where
+    f v p = map (\(k, g) -> (p <> k, v g)) pairs
+    pairs = [ ("-singleton", singletonChainGraph)
+            , ("-pair", pairChainGraph)
+            , ("-triangle", triangleChainGraph)
+            , ("-peterson", petersonChainGraph)
+            , ("-twenty", twentyChainGraph)
+            , ("-hoffman-singleton", hoffmanSingletonGraph)
+            ]
+
+prettyVersions :: HM.HashMap ChainwebVersion T.Text
+prettyVersions = HM.fromList . map swap $ HM.toList chainwebVersions
+
+chainwebVersionIds :: HM.HashMap Word32 ChainwebVersion
+chainwebVersionIds =
+    HM.fromList . map (chainwebVersionId &&& id) $ HM.elems chainwebVersions
+
 -- -------------------------------------------------------------------------- --
 -- Test instances
 --
 -- The code in this section must not be called in production.
-
--- For all production instances of Chainweb, including test nets, the
--- 'ChainwebVersion' is a constant constructor that statically determines all
--- parameters of that version.
---
--- For testing instances, however, we require the oppportunity to uses different
--- parameters. Defining a new static 'ChainwebVersion' value for each test
--- setting would create too much overhead and polute the code base. Instead we
--- parameterize the respective 'ChainwebVersion' constructors with for defining
--- the dynamically configurable parameters.
---
--- When deserializing a parameterized test 'ChainwebVersion' we need a way to
--- restore the dynamic parameters. For that hash those parameters (with 15bit
--- precision) and store the respective parameters in a global hash table. This
--- hashtable is used only for testing and must never be used by prodcution code.
 --
-type TestChainwebVersionMap = HM.HashMap Word32 ChainwebVersion
-
--- | Global map for keeping track of Test Chainweb Versions with non-static
--- parameters.
---
-testChainwebVersionMap :: TVar TestChainwebVersionMap
-testChainwebVersionMap = unsafePerformIO $ newTVarIO mempty
-{-# NOINLINE testChainwebVersionMap #-}
 
 toTestChainwebVersion :: HasCallStack => ChainwebVersion -> Word32 -> Word32
-toTestChainwebVersion Testnet00 _
-    = error "toTestChainwebVersion must not be called for a production isntances"
-toTestChainwebVersion v i = unsafePerformIO $ do
-    m <- readTVarIO testChainwebVersionMap
-    case HM.lookup h m of
-        Just _ -> return ()
-        Nothing -> atomically
-            $ modifyTVar' testChainwebVersionMap $ HM.insert h v
-    return h
-  where
-    h = i .|. (testChainwebVersionMask .&. int (hash v))
+toTestChainwebVersion Testnet00 _ =
+    error "toTestChainwebVersion must not be called for a production instance"
+toTestChainwebVersion v i = i .|. (testChainwebVersionMask .&. int (hash v))
```

Comment:
I propose that we don't use the `hash` function any more but just hard-code the version-id component for each know graph.

Reply:
That should give us a small perf boost :+1: